### PR TITLE
Budi analyzes entities

### PR DESCRIPTION
**Pull request by RazvanBudacaTechquarter**
Sending a POST request to the servlet with the lyrics as parameter will return a list of the most salient entities, each having a name and a salience score.

Sensitive data will be hidden later with Olya, after the branches merge. 
Tests make the PRs look huge. Maybe I should separate them? If you have any test suggestions, feel free to message me.

### CODE REVIEW

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (jeinygroove) [inline src/main/java/com/google/alpollo/AnalysisHelper.java]: remove empty line between imports

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (jeinygroove) [inline src/main/java/com/google/alpollo/AnalysisHelper.java]: add spaces in math operations

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (jeinygroove) [inline src/main/java/com/google/alpollo/AnalysisHelper.java]: it's better to specify what part of information do we get

File: src/test/java/com/google/alpollo/AnalysisTest.java
Comment (jeinygroove) [inline src/test/java/com/google/alpollo/AnalysisTest.java]: remove empty line between imports

File: src/test/java/com/google/alpollo/AnalysisTest.java
Comment (jeinygroove) [inline src/test/java/com/google/alpollo/AnalysisTest.java]: remove empty line between imports

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (jeinygroove) [inline src/main/java/com/google/alpollo/AnalysisHelper.java]: I am sure that there must be a simple function for this action. I know that Scala has `.slice(10)`, for example, which takes first 10 elements of a list or all of them, it the length of the list is `<10`.

File: src/main/java/com/google/alpollo/EntityServlet.java
Comment (jeinygroove) [inline src/main/java/com/google/alpollo/EntityServlet.java]: remove empty line between imports

File: src/main/java/com/google/alpollo/EntityServlet.java
Comment (jeinygroove) [inline src/main/java/com/google/alpollo/EntityServlet.java]: remove empty line between imports

File: src/main/java/com/google/alpollo/EntityServlet.java
Comment (jeinygroove) [inline src/main/java/com/google/alpollo/EntityServlet.java]: I think that this comment can be deleted, we simplify entities not because it will be more comfortable to work with them on the client side, but because we don't need all the information which entity contains, that's why there's no need to send from servlet a lot of useless and more heavy data.

File: src/main/java/com/google/alpollo/model/SongEntity.java
Comment (jeinygroove) [inline src/main/java/com/google/alpollo/model/SongEntity.java]: It would be great to specify the order in the name of the variable

Comment (jeinygroove) [reply src/main/java/com/google/alpollo/model/SongEntity.java]: Also maybe it's better to make it descending, because you only use it with `reverseOrder`

File: src/test/java/com/google/alpollo/AnalysisTest.java
Comment (jeinygroove) [inline src/test/java/com/google/alpollo/AnalysisTest.java:59]: I'm not sure that we need to write tests which depends on the other API, because what if API will change and it'll send more than 10 entities in this text or it'll change an algorithm for defining salience of the word.

I suggest to emulate the result of the API, for example you can manually define `entityList`, so you'll check only your own methods.

File: src/test/java/com/google/alpollo/AnalysisTest.java
Comment (jeinygroove) [inline src/test/java/com/google/alpollo/AnalysisTest.java]: I'm not sure that we need to write tests which depends on the other API, because what if API will change and it'll send more than 10 entities in this text or it'll change an algorithm for defining salience of the word.

I suggest to emulate the result of the API, for example you can manually define entityList, so you'll check only your own methods.

File: src/test/java/com/google/alpollo/AnalysisTest.java
Comment (jeinygroove) [inline src/test/java/com/google/alpollo/AnalysisTest.java]: you call only static methods of the `AnalysisHelper`, so you don't need this variable, just call them with the class name

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (jeinygroove) [inline src/main/java/com/google/alpollo/AnalysisHelper.java:2]: Now I see that all methods in this class are `static`, so it's good to make it stateless

This's called Utility class and you can find a lot of best practises for it, for example, [here](https://stackoverflow.com/questions/31409982/java-best-practice-class-with-only-static-methods)

File: src/test/java/com/google/alpollo/AnalysisTest.java
Comment (jeinygroove) [inline src/test/java/com/google/alpollo/AnalysisTest.java:73]: Please, add comment to this test

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (RazvanBudacaTechquarter) [reply src/main/java/com/google/alpollo/AnalysisHelper.java]: I don't think there's any method like that in Java. I'll give it a search.

File: src/test/java/com/google/alpollo/AnalysisTest.java
Comment (RazvanBudacaTechquarter) [reply src/test/java/com/google/alpollo/AnalysisTest.java:59]: Initially I added everything because I wanted to check if everything works correctly. I could make a separate test to see if `getEntityList()` works and use a manually defined list here.
Let's see what the guys have to say about it too.

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (jeinygroove) [reply src/main/java/com/google/alpollo/AnalysisHelper.java]: [Look what I've found](https://stackoverflow.com/questions/11001720/get-only-part-of-an-array-in-java)

Comment (RazvanBudacaTechquarter) [reply src/main/java/com/google/alpollo/AnalysisHelper.java]: Apparently if the old array has less elements than you request, it will fill the new one with nulls.

Comment (jeinygroove) [reply src/main/java/com/google/alpollo/AnalysisHelper.java]: yes, but for `to` parameter you can make `int to = min(10, arr.size())`, problem solved)

Comment (jeinygroove) [reply src/main/java/com/google/alpollo/AnalysisHelper.java]: Also avoid "magic constants", move `10` to the constant variable

**Review by jeinygroove — COMMENTED** [review]
I merged your branch with config changes.

Please, don't forget regularly upload new changes to your branch and fix conflicts.

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (RazvanBudacaTechquarter) [reply src/main/java/com/google/alpollo/AnalysisHelper.java]: Tried using the array methods like this:
`List<SongEntity> topSalientEntities = Arrays.asList(Arrays.copyOfRange(simplifiedEntityList.toArray(), 0, 9));`

However, when converting back to list, it contains only simple objects, not SongEntities.

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (jeinygroove) [inline src/main/java/com/google/alpollo/AnalysisHelper.java]: `MAX_ENTITIES` instead of `10`

**Review by jeinygroove — APPROVED** [review]

**Comment by jeinygroove** [comment]
@BudiGabe please, name commits according to the code, so we'll be able to understand what was done without opening it

**Comment by RazvanBudacaTechquarter** [comment]
Okok no more jokes in commits :))

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (jgaillard85) [inline src/main/java/com/google/alpollo/AnalysisHelper.java]: Constants are static and final

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (jgaillard85) [inline src/main/java/com/google/alpollo/AnalysisHelper.java]: The name of this method is misleading because it looks like a getter method but does not return anything.

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (jgaillard85) [inline src/main/java/com/google/alpollo/AnalysisHelper.java:90]: What is the point of this operation?

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (jgaillard85) [inline src/main/java/com/google/alpollo/AnalysisHelper.java]: Only create this object if needed, so put it in the else branch

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (jgaillard85) [inline src/main/java/com/google/alpollo/AnalysisHelper.java]: You could use the sublist method to avoid creating a new list object for this, if you don't need a copy

File: src/main/java/com/google/alpollo/EntityServlet.java
Comment (jgaillard85) [inline src/main/java/com/google/alpollo/EntityServlet.java]: This is not a constant, so the field name should be lower case

Comment (jgaillard85) [reply src/main/java/com/google/alpollo/EntityServlet.java]: This doesn't even need to be a field, it could be a local variable

File: src/main/java/com/google/alpollo/EntityServlet.java
Comment (jgaillard85) [inline src/main/java/com/google/alpollo/EntityServlet.java]: Why do you create a copy of the list returned by getEntityList here?

File: src/main/java/com/google/alpollo/EntityServlet.java
Comment (jgaillard85) [inline src/main/java/com/google/alpollo/EntityServlet.java]: Maybe this sorting should happen in the getTopSalientEntities, as the method name suggests it returns the top ones, while really at the moment it returns the first ones

File: src/test/java/com/google/alpollo/AnalysisTest.java
Comment (jgaillard85) [reply src/test/java/com/google/alpollo/AnalysisTest.java:59]: I think you should have tests that test that your own code is doing what it is supposed to do, and also tests that make sure that the API you are relying on is still doing what you think it is doing. It is generally better to have those separate, so that you immediately know if something breaks whether it is something you did or it is related to a change in the API

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (RazvanBudacaTechquarter) [reply src/main/java/com/google/alpollo/AnalysisHelper.java:90]: I used this to format the double to always show 2 decimal places. It was pretty ugly with all those decimals.

File: src/main/java/com/google/alpollo/EntityServlet.java
Comment (RazvanBudacaTechquarter) [reply src/main/java/com/google/alpollo/EntityServlet.java]: Ohh that's my bad, apparently I forgot some stuff after merging. Thanks for the heads up,

File: src/test/java/com/google/alpollo/AnalysisTest.java
Comment (RazvanBudacaTechquarter) [reply src/test/java/com/google/alpollo/AnalysisTest.java:59]: Should I leave that to a separate PR, as it is already pretty huge?

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (amaurymedeiros) [reply src/main/java/com/google/alpollo/AnalysisHelper.java:90]: You can store the value as it is. When displaying that to the user, i.e. printing to the UI, you can format the string to display only 2 decimal places

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (amaurymedeiros) [inline src/main/java/com/google/alpollo/AnalysisHelper.java]: This modifies the list passed as an argument. We shouldn't do that as it's not expected by looking at the method signature. You should make a copy instead.

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (amaurymedeiros) [inline src/main/java/com/google/alpollo/AnalysisHelper.java]: nit: you can merge these lines, e.g. `return simplifiedEntityList.subList(FIRST_ENTITY, MAX_ENTITIES);`

File: src/main/java/com/google/alpollo/EntityServlet.java
Comment (amaurymedeiros) [inline src/main/java/com/google/alpollo/EntityServlet.java]: Document this constant saying it's a parameter from requests to the servlet, etc.

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (amaurymedeiros) [inline src/main/java/com/google/alpollo/AnalysisHelper.java]: Please document these constants

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (RazvanBudacaTechquarter) [reply src/main/java/com/google/alpollo/AnalysisHelper.java:90]: Yes, it can be fixed while printing the UI, but without rounding I can't really write tests for it. The online API only shows me the first 2 decimals, so the only way to compare them was to round the numbers.

File: src/main/java/com/google/alpollo/AnalysisHelper.java
Comment (jgaillard85) [inline src/main/java/com/google/alpollo/AnalysisHelper.java]: Here I think you want a sublist of tempList.

File: src/main/java/com/google/alpollo/EntityServlet.java
Comment (jgaillard85) [inline src/main/java/com/google/alpollo/EntityServlet.java]: For it to be a constant, you have to make it final

File: src/test/java/com/google/alpollo/AnalysisTest.java
Comment (jgaillard85) [reply src/test/java/com/google/alpollo/AnalysisTest.java:59]: You can leave it as is

**Review by jgaillard85 — APPROVED** [review]

**Pull request merged**
